### PR TITLE
Add proliant xl420

### PR DESCRIPTION
**Pull request by ibrahimkk-moideen**
Adding support for HP Proliant XL420 Server model #33 

**Test results:**
```
# HELP up Was the last scrape of chassis monitor successful.
# TYPE up gauge
up 1
# HELP xl420_memory_status Current memory status 1 = OK, 0 = BAD
# TYPE xl420_memory_status gauge
xl420_memory_status{totalSystemMemoryGiB="128"} 1
# HELP xl420_physical_drive_status Current physical drive status 1 = OK, 0 = BAD, -1 = DISABLED
# TYPE xl420_physical_drive_status gauge
xl420_physical_drive_status{id="16",location="Redacted",name="HpSmartStorageDiskDrive",serialnumber="Redacted"} 1
xl420_physical_drive_status{id="16",location="Redacted",name="HpSmartStorageDiskDrive",serialnumber="Redacted"} 1
[***output truncated***]
xl420_physical_drive_status{id="27",location="Redacted",name="HpSmartStorageDiskDrive",serialnumber="Redacted"} 1
xl420_physical_drive_status{id="27",location="Redacted",name="HpSmartStorageDiskDrive",serialnumber="Redacted"} 1
# HELP xl420_power_supply_output Power supply output in watts
# TYPE xl420_power_supply_output gauge
xl420_power_supply_output{memberId="1",sparePartNumber="754381-001"} 172
xl420_power_supply_output{memberId="2",sparePartNumber="754381-001"} 238
# HELP xl420_power_supply_status Current power supply status 1 = OK, 0 = BAD
# TYPE xl420_power_supply_status gauge
xl420_power_supply_status{memberId="1",sparePartNumber="754381-001"} 1
xl420_power_supply_status{memberId="2",sparePartNumber="754381-001"} 1
# HELP xl420_power_supply_total_capacity Total output capacity of all the power supplies
# TYPE xl420_power_supply_total_capacity gauge
xl420_power_supply_total_capacity{memberId="0"} 1600
# HELP xl420_power_supply_total_consumed Total output of all power supplies in watts
# TYPE xl420_power_supply_total_consumed gauge
xl420_power_supply_total_consumed{memberId="0"} 410
# HELP xl420_thermal_fan_speed Current fan speed in the unit of percentage, possible values are 0 - 100
# TYPE xl420_thermal_fan_speed gauge
xl420_thermal_fan_speed{name="Fan 1"} 39
xl420_thermal_fan_speed{name="Fan 10"} 51
xl420_thermal_fan_speed{name="Fan 2"} 64
xl420_thermal_fan_speed{name="Fan 3"} 64
xl420_thermal_fan_speed{name="Fan 4"} 61
xl420_thermal_fan_speed{name="Fan 5"} 51
xl420_thermal_fan_speed{name="Fan 6"} 39
xl420_thermal_fan_speed{name="Fan 7"} 64
xl420_thermal_fan_speed{name="Fan 8"} 64
xl420_thermal_fan_speed{name="Fan 9"} 61
# HELP xl420_thermal_fan_status Current fan status 1 = OK, 0 = BAD
# TYPE xl420_thermal_fan_status gauge
xl420_thermal_fan_status{name="Fan 1"} 1
xl420_thermal_fan_status{name="Fan 10"} 1
xl420_thermal_fan_status{name="Fan 2"} 1
xl420_thermal_fan_status{name="Fan 3"} 1
xl420_thermal_fan_status{name="Fan 4"} 1
xl420_thermal_fan_status{name="Fan 5"} 1
xl420_thermal_fan_status{name="Fan 6"} 1
xl420_thermal_fan_status{name="Fan 7"} 1
xl420_thermal_fan_status{name="Fan 8"} 1
xl420_thermal_fan_status{name="Fan 9"} 1
# HELP xl420_thermal_sensor_status Current sensor status 1 = OK, 0 = BAD
# TYPE xl420_thermal_sensor_status gauge
xl420_thermal_sensor_status{name="01-Front Ambient"} 1
xl420_thermal_sensor_status{name="02-CPU 1"} 1
[***output truncated***]
xl420_thermal_sensor_status{name="46-HD Board"} 1
xl420_thermal_sensor_status{name="48-HD Board"} 1
# HELP xl420_thermal_sensor_temperature Current sensor temperature reading in Celsius
# TYPE xl420_thermal_sensor_temperature gauge
xl420_thermal_sensor_temperature{name="01-Front Ambient"} 19
xl420_thermal_sensor_temperature{name="02-CPU 1"} 40
[***output truncated***]
xl420_thermal_sensor_temperature{name="46-HD Board"} 32
xl420_thermal_sensor_temperature{name="48-HD Board"} 29
```

Adding support for HP DL360 Redfish responses from iLO4. #34 

**Test results:**
```
# HELP dl360_disk_drive_status Current Disk Drive status 1 = OK, 0 = BAD
# TYPE dl360_disk_drive_status gauge
dl360_disk_drive_status{Id="0",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="1",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="2",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="3",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="4",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="5",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="6",location="Redacted",name="HpSmartStorageDiskDrive"} 1
dl360_disk_drive_status{Id="7",location="Redacted",name="HpSmartStorageDiskDrive"} 1
# HELP dl360_logical_drive_raid Current Logical Drive Raid
# TYPE dl360_logical_drive_raid gauge
dl360_logical_drive_raid{logicaldrivename="Redacted",name="HpSmartStorageLogicalDrive",raid="0",volumeuniqueidentifier="Redacted"} 1
# HELP dl360_power_supply_output Power supply output in watts
# TYPE dl360_power_supply_output gauge
dl360_power_supply_output{memberId="1",sparePartNumber="754377-001"} 116
dl360_power_supply_output{memberId="2",sparePartNumber="754377-001"} 0
# HELP dl360_power_supply_status Current power supply status 1 = OK, 0 = BAD
# TYPE dl360_power_supply_status gauge
dl360_power_supply_status{memberId="1",sparePartNumber="754377-001"} 1
dl360_power_supply_status{memberId="2",sparePartNumber="754377-001"} 1
# HELP dl360_power_supply_total_capacity Total output capacity of all the power supplies
# TYPE dl360_power_supply_total_capacity gauge
dl360_power_supply_total_capacity{memberId="0"} 500
# HELP dl360_power_supply_total_consumed Total output of all power supplies in watts
# TYPE dl360_power_supply_total_consumed gauge
dl360_power_supply_total_consumed{memberId="0"} 116
# HELP dl360_thermal_fan_speed Current fan speed in the unit of percentage, possible values are 0 - 100
# TYPE dl360_thermal_fan_speed gauge
dl360_thermal_fan_speed{name="Fan 1"} 19
dl360_thermal_fan_speed{name="Fan 2"} 15
dl360_thermal_fan_speed{name="Fan 3"} 19
dl360_thermal_fan_speed{name="Fan 4"} 19
dl360_thermal_fan_speed{name="Fan 5"} 21
dl360_thermal_fan_speed{name="Fan 6"} 27
dl360_thermal_fan_speed{name="Fan 7"} 27
# HELP dl360_thermal_fan_status Current fan status 1 = OK, 0 = BAD
# TYPE dl360_thermal_fan_status gauge
dl360_thermal_fan_status{name="Fan 1"} 1
dl360_thermal_fan_status{name="Fan 2"} 1
dl360_thermal_fan_status{name="Fan 3"} 1
dl360_thermal_fan_status{name="Fan 4"} 1
dl360_thermal_fan_status{name="Fan 5"} 1
dl360_thermal_fan_status{name="Fan 6"} 1
dl360_thermal_fan_status{name="Fan 7"} 1
```

### CODE REVIEW

**Review by derrick-dacosta — CHANGES_REQUESTED** [review]

File: hpe/dl360/power.go [hunk 1]
```diff
@@ -63,6 +63,7 @@ type PowerSupply struct {
 // OemPower is the top level json object for historical data for wattage
 type OemPower struct {
 	Hpe Hpe `json:"Hpe"`
```
[the remark this answers is not in the document]
Comment (ibrahimkk-moideen) [reply hpe/dl360/power.go]: Updated in [802c769](https://github.com/Comcast/fishymetrics/pull/35/commits/802c76924269b508d053d8692fa519cb3d124d60).

File: hpe/xl420/drive.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	} `json:"Members"`
+	MembersCount int `json:"Members@odata.count,omitempty"`
+	Links        struct {
+		LogicalDrives struct {
+			URL string `json:"href"`
+		} `json:"LogicalDrives,omitempty"`
+		PhysicalDrives struct {
+			URL string `json:"href"`
+		} `json:"PhysicalDrives,omitempty"`
+	}
```
[the remark this answers is not in the document]
Comment (ibrahimkk-moideen) [reply hpe/xl420/drive.go]: Updated in [802c769](https://github.com/Comcast/fishymetrics/pull/35/commits/802c76924269b508d053d8692fa519cb3d124d60).

File: hpe/dl360/exporter.go [hunk 3]
```diff
@@ -169,6 +169,21 @@ func NewExporter(ctx context.Context, target, uri, profile string) *Exporter {
 						logicalDriveURLs = append(logicalDriveURLs, member.URL)
 					}
 				}
+			} else {
```
[the remark this answers is not in the document]
Comment (ibrahimkk-moideen) [reply hpe/dl360/exporter.go]: Updated in [802c769](https://github.com/Comcast/fishymetrics/pull/35/commits/802c76924269b508d053d8692fa519cb3d124d60).

File: hpe/dl360/exporter.go [hunk 4]
```diff
@@ -184,6 +199,18 @@ func NewExporter(ctx context.Context, target, uri, profile string) *Exporter {
 						physicalDriveURLs = append(physicalDriveURLs, member.URL)
 					}
 				}
+			} else {
```
[the remark this answers is not in the document]
Comment (ibrahimkk-moideen) [reply hpe/dl360/exporter.go]: Updated in [802c769](https://github.com/Comcast/fishymetrics/pull/35/commits/802c76924269b508d053d8692fa519cb3d124d60).

File: hpe/xl420/exporter.go [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+
+	if arrayControllers.MembersCount > 0 {
+		for _, controller := range arrayControllers.Members {
+			getController, err := getDriveEndpoints(fqdn.String()+controller.URL, target, retryClient)
+			if err != nil {
+				log.Error("error when getting array controller from "+XL420, zap.Error(err), zap.Any("trace_id", ctx.Value("traceID")))
+				return nil
+			}
+
+			if getController.Links.LogicalDrives.URL != "" {
```
[the remark this answers is not in the document]
Comment (ibrahimkk-moideen) [reply hpe/xl420/exporter.go]: Updated in [802c769](https://github.com/Comcast/fishymetrics/pull/35/commits/802c76924269b508d053d8692fa519cb3d124d60).

File: hpe/xl420/exporter.go [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+
+				if logicalDrives.MembersCount > 0 {
+					for _, logicalDrive := range logicalDrives.Members {
+						tasks = append(tasks,
+							pool.NewTask(common.Fetch(fqdn.String()+logicalDrive.URL, LOGICALDRIVE, target, profile, retryClient)))
+					}
+				}
+			}
+
+			if getController.Links.PhysicalDrives.URL != "" {
```
[the remark this answers is not in the document]
Comment (ibrahimkk-moideen) [reply hpe/xl420/exporter.go]: Updated in [802c769](https://github.com/Comcast/fishymetrics/pull/35/commits/802c76924269b508d053d8692fa519cb3d124d60).

**Review by derrick-dacosta — APPROVED** [review]
LGTM!

**Pull request merged**